llvm/unittests/ADT/AtomicUniquePointerTest.cpp

```diff
@@ -0,0 +1,75 @@
+//===- llvm/unittests/ADT/AtomicUniquePointerTest.cpp ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/AtomicUniquePointer.h"
+#include "llvm/Support/ThreadPool.h"
+#include "gtest/gtest.h"
+#include <memory>
+
+using namespace llvm;
+
+namespace {
+
+TEST(AtomicUniquePointerTest, Create) {
+  AtomicUniquePointer<int> ImplPtr;
+  SmallVector<int*, 4> Results(4);
+
+  auto createImpl = [&ImplPtr, &Results](unsigned Idx) {
+    auto Impl = std::make_unique<int>();
+    int *ExistingImpl = nullptr;
+    (void)ImplPtr.compare_exchange_strong(ExistingImpl, Impl);
+    Results[Idx] = ImplPtr.load();
+  };
+
+  ThreadPool Threads;
+  for (unsigned I = 0; I < 4; ++I)
+    Threads.async(createImpl, I);
+  Threads.wait();
+
+  EXPECT_NE(ImplPtr.load(), nullptr);
+  
+  // All the threads should get the same results.
+  SmallVector<int*, 4> ExpectedPtrs(4, ImplPtr.load());
+  EXPECT_EQ(Results, ExpectedPtrs);
+}
+
+
+TEST(AtomicUniquePointerTest, LinkList) {
+  // Test atomic insertion. Insert node S as Root -> S -> Root.Next.
+  struct DummyAtomicList {
+    AtomicUniquePointer<DummyAtomicList> Next;
+  };
+  DummyAtomicList Root;
+  auto insertNode = [&Root]() {
+    auto S = std::make_unique<DummyAtomicList>();
+    DummyAtomicList *CurrentHead = nullptr;
```

Comment:
Why is this a raw pointer instead of `std::unique_ptr<>`?

Reply:
I copy paste from real usecase here: https://github.com/apple/llvm-project/blob/experimental/cas/main/llvm/lib/CAS/HashMappedTrie.cpp#L175

I didn't think too much about lifetime.

Reply:
I guess we need a `T*&` as the parameter for atomic.compare_exchange.